src/v0/destinations/ga4/utils.js

```diff
@@ -336,30 +313,85 @@ const validateEventName = (event) => {
   }
 
   if (isReservedWebCustomPrefixName(event)) {
-    throw new InstrumentationError(
-      'Reserved custom prefix names are not allowed',
-    );
+    throw new InstrumentationError('Reserved custom prefix names are not allowed');
   }
 
-  if (!isEventNameValid(event)){
+  if (!isEventNameValid(event)) {
     throw new InstrumentationError(
       'Event name should only contain letters, numbers, and underscores and event name must start with a letter',
     );
   }
-}
+};
+
+/**
+ * Function to verify user_property value type
+ * @param {*} value
+ * @returns
+ */
+const isValidValueType = (value) =>
+  (typeof value === 'string' && value.trim() !== '') ||
+  typeof value === 'number' ||
+  typeof value === 'boolean';
+
+/**
+ * Function to validate user_property name and it's value
+ * user_property name should start with alphabetic characters
+ * user_property name length should not more then 24 characters
+ * user_property name should not start with reserved prefixes
+ * user_property value should not null and empty
+ * user_property value type should either string, number or boolean
+ * Ref - https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference?client_type=gtag#reserved_parameter_names
+ * @param {*} name
+ * @returns
+ */
+const isValidUserProperty = (key, value) => {
+  const namePattern = /^[A-Za-z]\w{0,23}$/;
+  const reservedPrefixesNames = ['google_', 'ga_', 'firebase_'];
+
+  const isValidKey =
+    namePattern.test(key) &&
+    !reservedPrefixesNames.some((prefix) => key.toLowerCase().startsWith(prefix));
+
+  const isValidValue = isDefinedAndNotNull(value) && isValidValueType(value);
+
+  return isValidKey && isValidValue;
+};
+
+/**
+ * Function to validate and prepare user_properties
+ * @param {*} message
+ */
+const prepareUserProperties = (message) => {
```

Comment:
Good refactoring, now write unit test for `prepareUserProperties` in file `util.test.js` --> which makes easy & convenient to cover varied test cases

Reply:
you can focus on test case payloads only with combination of 2 fields 'properties.user_properties', 'context.traits'